Improve SA1311 code fix

pdelvo opened the pull request:
Implements the changes discussed in #469.

No tests are implemented because we have not yet fully agreed on the solution.

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1311CodeFixProvider.cs:
```diff
@@ -43,8 +45,69 @@ public override ImmutableArray<string> GetFixableDiagnosticIds()
+
+        private bool IsValidNewMemberName(INamedTypeSymbol typeSymbol, string name)
+        {
+            foreach (var member in typeSymbol.GetMembers())
```
`typeSymbol.GetMembers(name)`

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1311CodeFixProvider.cs:
```diff
@@ -43,8 +45,69 @@ public override ImmutableArray<string> GetFixableDiagnosticIds()
+            {
+                if (member.Kind == SymbolKind.Method)
+                {
+                    // Field names can't colide with method names
```
:exclamation: A class is not allowed to have a field and a method with the same name.

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1311CodeFixProvider.cs:
```diff
@@ -43,8 +45,69 @@ public override ImmutableArray<string> GetFixableDiagnosticIds()
+            }
+            else
+            {
+                return this.IsValidNewMemberName(typeSymbol.BaseType, name);
```
There is no conflict if a field in a derived class has the same name as a member of the base class.

pdelvo replied:
If there is a member declared in a base class and the field is getting the same name you will access that field instead of the base class member. Thats a breaking change

sharwell replied:
> the renamer is smart...

Yes, yes it really, really is! :sunglasses:

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1311CodeFixProvider.cs:
```diff
@@ -43,8 +45,69 @@ public override ImmutableArray<string> GetFixableDiagnosticIds()
+                        return;
+                    }
+
+                    while (!this.IsValidNewMemberName(typeDeclarationSymbol, newName) || newName == typeSyntax.GetTypeName())
```
I don't think we need a `while` loop. Further conflicts would be addressed by #436, #437, etc.

sharwell replied:
:bulb: I do not think the test against `typeSyntax.GetTypeName()` is necessary. However, if it is then I would move it into `IsValidNewMemberName`.

The reason I don't think it's needed is we are making a very special case here for when a field is specifically used and named in order for it to be a backing field for a property (or property-like method). In any other case, the `someName` &rarr; `SomeNameValue` may or may not make sense, and I would prefer to leave this determination for the user.

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1311CodeFixProvider.cs:
```diff
@@ -43,8 +45,69 @@ public override ImmutableArray<string> GetFixableDiagnosticIds()
+        private TypeDeclarationSyntax GetParentTypeDeclaration(SyntaxToken token)
+        {
+            SyntaxNode parent = token.Parent;
+            while (!(parent is TypeDeclarationSyntax))
```
:question: Can you replace the `is`/`cast` here with `as`?

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers/Helpers/NamedTypeHelpers.cs:
```diff
@@ -69,5 +69,23 @@ internal static bool IsContainedInNativeMethodsClass(SyntaxNode syntax)
             return false;
         }
+
+        internal static string GetTypeName(this TypeDeclarationSyntax typeDeclaration)
```
:bulb: Perhaps leave this method for another time?

pdelvo commented:
I applied the changes you suggested

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1311CodeFixProvider.cs:
```diff
@@ -43,8 +47,58 @@ public override ImmutableArray<string> GetFixableDiagnosticIds()
+                        return;
+                    }
+
+                    while (!this.IsValidNewMemberName(typeDeclarationSymbol, newName) || newName == typeDeclarationSymbol.Name)
```
Can you change this line to the following?

```csharp
if (!this.IsValidNewMemberName(typeDeclarationSymbol, newName))
```

pdelvo replied:
I agree with putting the name check into the IsValidNewMemberName method. But without the while the code fix may suggest a code fix that will not compile if there are still name conflicts after the fix.

sharwell replied:
It's OK for it to propose a name that conflicts. It just means the user is responsible for choosing an appropriate name.

The user experience surrounding conflicting names is unpleasant, but that's another issue that will likely need to wait for a change in Roslyn.

sharwell replied:
@pdelvo As a further clarification, I am not opposed to expanding the algorithm for this code fix in the future, but I think each move in that direction should be narrowly scoped and designed to address a particularly common scenario. For this pull request, the only situation I'm hoping to cover is the simplest case where a backing field and property differ only by the casing of the first letter. :+1:

sharwell commented:
With the one remaining change I suggested in place, I think we are good to go on the implementation :+1:

sharwell commented:
:bulb: You might consider enabling the batch fix all provider for this. I saved a great deal of time creating #491 by enabling the fix all provider beforehand.

pdelvo commented:
I implemented your additional feedback, rebased everything on master and implemented tests. I have no idea why this one "contains merge conflicts" because merging this into master would be a fast forward.
